File: Data Structures & Algorithms/partition-equal-subset-sum/submission-6.java
class Solution {
    Boolean[][] dp;
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int x : nums)
            sum += x;
        if(sum%2 != 0) return false;
        sum /= 2;
        dp = new Boolean[nums.length][sum+1];
        return helper(nums, 0, 0, sum);
    }
    public boolean helper(int[] nums, int i, int curr, int sum) {
        if(curr==sum) {
            return true;
        }
        if(curr>sum) {
            return false;
        }
        if(i==nums.length) return curr==sum;
        if(dp[i][curr]!=null) return dp[i][curr];
        dp[i][curr] = helper(nums, i+1, curr+nums[i], sum)|| helper(nums, i+1, curr, sum);
        return dp[i][curr];
    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-6.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n==1:
            return nums[0]

        def helper(l: int, r: int) -> int:
            prev = 0
            curr = nums[l]
            for i in range(l+1, r+1):
                r = nums[i]+prev
                nr = max(prev,curr)
                prev = curr
                curr = max(r,nr)
            return curr

        return max(helper(0,n-2), helper(1,n-1))